src/main/java/com/example/cinematicketbookingapp/service/ReservationService.java

```diff
@@ -0,0 +1,159 @@
+package com.example.cinematicketbookingapp.service;
+
+import com.example.cinematicketbookingapp.config.AppFunctionalValuesConstants;
+import com.example.cinematicketbookingapp.dto.ReservationCreationDataDto;
+import com.example.cinematicketbookingapp.dto.ReservationSummaryDto;
+import com.example.cinematicketbookingapp.exceptions.ReservationSystemClosedException;
+import com.example.cinematicketbookingapp.exceptions.ReservedSeatsNumberUnequalToTicketsNumberException;
+import com.example.cinematicketbookingapp.exceptions.SeatAlreadyReservedException;
+import com.example.cinematicketbookingapp.exceptions.SingleUnreservedSeatLeftException;
+import com.example.cinematicketbookingapp.mapper.ReservationDtoMapper;
+import com.example.cinematicketbookingapp.model.Reservation;
+import com.example.cinematicketbookingapp.model.Screening;
+import com.example.cinematicketbookingapp.model.Seat;
+import com.example.cinematicketbookingapp.model.TicketType;
+import com.example.cinematicketbookingapp.repository.ReservationRepository;
+import com.example.cinematicketbookingapp.repository.SeatRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+
+@Service
+@RequiredArgsConstructor
+public class ReservationService {
+
+    private final ReservationRepository reservationRepository;
+    private final ReservationDtoMapper reservationDtoMapper;
+    private final SeatRepository seatRepository;
+
+    @Transactional
+    public ReservationSummaryDto createReservation(ReservationCreationDataDto reservationCreationDataDto) {
+        Reservation reservation = reservationDtoMapper.mapToReservation(reservationCreationDataDto);
+
+        checkIfReservationDateTimeIsBeforeScreeningReservationSystemClosureDateTime(reservation.getScreening());
+        checkIfReservedSeatsAmountIsEqualToTicketsAmount(reservationCreationDataDto);
+        checkIfChosenSeatsAreNotAlreadyReserved(reservation.getSeats(), reservationCreationDataDto.screeningId());
+        checkIfUnreservedSingleSeatIsLeft(reservation.getSeats(), reservationCreationDataDto.screeningId());
+
+        return reservationDtoMapper.mapToReservationSummaryDto(reservationRepository.save(reservation)).toBuilder()
```

Comment:
Ale głęboko ukryty ten save. Wyciągnąłbym żeby był bardziej widoczny.

Reply:
Rozumiem, że tutaj kwestią jest czytelność kodu. Czyli lepiej byłoby przypisać rezultat metody `save()` do zmiennej i dopiero później przemapować tę zmienną zwracając rezultat mapowania przy okazji?

Reply:
Tak, ja bym tak zrobił. To jest jednak kluczowe wywołanie w tej funkcji, a jest schowane.